addon/components/abstract-modal.js

```diff
@@ -0,0 +1,146 @@
+/* global Ps */
+
+/**
+ * Functionality to style headder/footer elements depending on scroll state, content hidden under header/footer
+ */
+
+import Ember from 'ember'
+import $ from 'jquery'
+
+const {Component} = Ember
+
+export default Component.extend({
+  scrollBindingsSet: false,
+  containerObserver: null,  // @type {window.MutationObserver}
+  $containerEl: null, // @type {JQuery object}
+  $headerEl: null, // @type {JQuery object}
+  $footerEl: null, // @type {JQuery object}
+  containerElClassname: 'ps-container',
+  headerElClassname: 'input-header',
+  footerElClassname: 'actions',
+
+  /**
+   * create a window.MutationObserver on the container element
+   * mutation events callback invokes updateScrollStyles()
+   * @returns {MutationObserver} window.MutationObserver on container element
+   */
+
+  createMutationObserver () {
+    // create an observer instance
+    let mutationObserver = new window.MutationObserver((mutations) => {
+      mutations.forEach((mutation) => {
+        this.updateScrollStyles()
+      })
+    })
+
+    // configuration of the observer
+    const config = { attributes: false, childList: true, characterData: false, subtree: true }
+
+    // pass in the target DOM node, as well as the observer options
+    mutationObserver.observe(this.get('$containerEl').get(0), config)
+
+    return mutationObserver
+  },
+
+  /**
+  * update perfect-scrollbar before checking scoll position
+  * need to disconnect MutationObserver then re-establish to prevent infinite loop
+  * of mutation event triggers, callback updates, which triggers new mutation event...
+  */
+  updatePerfectScrollbar () {
+    let mutationObserver
+    const $containerEl = this.get('$containerEl')
+
+    Ps.update($containerEl.get(0))
+
+    if (this.get('containerObserver')) {
+      this.get('containerObserver').disconnect()
+      mutationObserver = this.createMutationObserver()
+    }
+
+    if (this.isDestroyed || this.isDestroying) {
+      return
+    } else {
+      this.set('containerObserver', mutationObserver)
+    }
+  },
+
+  /**
+  * Checks scroll position within container element to add/remove scroll styling to header/footer elements
+  */
+  updateScrollStyles () {
+    const $containerEl = this.get('$containerEl')
+    const $headerEl = this.get('$headerEl')
+    const $footerEl = this.get('$footerEl')
```

Comment:
We should create an addon that allows you to do something like below via destructuring:
```js
const [$containerEl, $headerEl, $footerEl] = this.getProps('$containerEl', '$headerEl', '$footerEl')
```